src/nc/gui/CxxDocument.cpp

```diff
@@ -51,7 +51,7 @@ namespace nc { namespace gui {
 
 CxxDocument::CxxDocument(QObject *parent, std::shared_ptr<const core::Context> context):
     QTextDocument(parent),
-    core::likec::CxxDocument(context),
+    core::likec::CxxDocument(context.get()),
```

Comment:
Not sure I understand the question.
The idea is:
1. gui::CxxDocument is the guy owning the context (keeping a shared pointer to it).
2. likec::CxxDocument is the guy not owning the document (not keeping any pointer to it).
3. As long as nobody destroys the context which is used by likec::CxxDocument, everything is fine.
4. Nobody destroys it in the GUI, because gui::CxxDocument owns the context.
5. Nobody destroys it in the command-line tool, because Context resides on the stack there.

Shared pointers to the context are actually passed around essentially in the GUI only: gui::Project stores a pointer to the last context where decompilation took or is taking place. Model objects (gui::InspectorModel, gui::CxxDocument) store a pointer to the last context where decompilation completed (because the user probably does not want to see an empty widget immediately after she pushes Decompile and until decompilation succeeds; also, it is a bit scary and undefined behavior to show something that is being modified in a background thread, the one doing decompilation).

A similar story with Instructions and Image: Context needs to own them (because they are needed, e.g., during decompilation), so, it keeps a shared pointer to them. Projects owns them, because it needs to set them in the Context before starting decompilation. Also, models of widgets that show instructions (InstructionsModel) and image-related stuff (SectionsModel, SymbolsModel) store shared pointers to Instructions resp. Image, because they cannot really continue working if the data they are providing disappears.

In the comand-line decompiler everything is simple: no shared ownership takes place, and I would prefer to keep it that way.